compiler/src/iree/compiler/Dialect/LinalgExt/IR/TilingInterfaceImpl.cpp

```diff
@@ -438,6 +439,164 @@ LogicalResult GatherOp::generateScalarImplementation(OpBuilder &b, Location loc,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// MapScatterOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<utils::IteratorType> MapScatterOp::getLoopIteratorTypes() {
+  SmallVector<utils::IteratorType> iteratorTypes(getInputRank(),
+                                                 utils::IteratorType::parallel);
+  return iteratorTypes;
+}
+
+SmallVector<Range> MapScatterOp::getIterationDomain(OpBuilder &builder) {
+  Location loc = getLoc();
+  OpFoldResult zero = builder.getIndexAttr(0);
+  OpFoldResult one = builder.getIndexAttr(1);
+  SmallVector<Range> ranges;
+  for (auto dim : llvm::seq<int64_t>(0, getInputRank())) {
+    OpFoldResult ub = getDim(builder, loc, getInput(), dim);
+    ranges.push_back(Range{zero, ub, one});
+  }
+  return ranges;
+}
+
+/// Generate the tiled implemenation of the map_scatter op, which scatters a
+/// slice of the input (taken from the input with the `offsets` and `sizes`)
+/// to the full output tensor, and inserts an addition of the `offsets` at the
+/// start of the transformation_region body.
+FailureOr<TilingResult>
+MapScatterOp::getTiledImplementation(OpBuilder &builder,
+                                     ArrayRef<OpFoldResult> offsets,
+                                     ArrayRef<OpFoldResult> sizes) {
+  assert(offsets.size() >= 1 && sizes.size() >= 1);
+  Location loc = getLoc();
+  SmallVector<Operation *> slices;
+
+  // Slice of the input.
+  auto oneAttr = builder.getI64IntegerAttr(1);
+  SmallVector<OpFoldResult> inputStrides(getInputRank(), oneAttr);
+  Operation *inputSlice =
+      getSlice(builder, loc, getInput(), offsets, sizes, inputStrides);
+  if (!inputSlice) {
+    return emitOpError("failed to get input slice");
+  }
+  Value tiledInput = inputSlice->getResult(0);
+  slices.push_back(inputSlice);
+
+  Operation *tiledOp = mlir::clone(builder, getOperation(), getResultTypes(),
+                                   {tiledInput, getOutput()});
+  auto tiledMapScatterOp = cast<MapScatterOp>(tiledOp);
+  auto indexTransformBuilder =
+      [&](ArrayRef<BlockArgument> srcIndices) -> SmallVector<Value> {
+    SmallVector<OpFoldResult> offsetIndices;
+    auto addMap = AffineMap::get(
+        2, 0, {builder.getAffineDimExpr(0) + builder.getAffineDimExpr(1)});
+    for (auto [srcIdx, offset] : llvm::zip_equal(srcIndices, offsets)) {
+      offsetIndices.push_back(affine::makeComposedFoldedAffineApply(
+          builder, loc, addMap, {OpFoldResult(srcIdx), offset}));
+    }
+    return getValueOrCreateConstantIndexOp(builder, loc, offsetIndices);
+  };
+  tiledMapScatterOp.insertTransformationAtStart(builder, indexTransformBuilder,
+                                                offsets.size());
```

Comment:
I think the logic can be simpler if we use IRMapping? The implementation in my mind is like

1. you use `cloneWithoutRegions` in the first place
2. create the region and block
3. generate the offset indices and do the mapping
4. clone all the ops.

It looks more straight-forward to me, and it is also easier for implementation comments. What do you think?

Reply:
Marked it resolved as I can see the value of the function, and we can reuse it.